pkl-core/src/main/java/org/pkl/core/packages/DependencyMetadata.java

```diff
@@ -423,8 +512,84 @@ private void write() throws IOException {
       if (dependencyMetadata.description != null) {
         jsonWriter.name("description").value(dependencyMetadata.description);
       }
+      jsonWriter.name("annotations");
+      writeAnnotations();
       jsonWriter.endObject();
       jsonWriter.close();
     }
+
+    private void writeAnnotations() throws IOException {
+      jsonWriter.beginArray();
+      for (var annotation : dependencyMetadata.annotations) {
+        writePObject(annotation);
+      }
+      jsonWriter.endArray();
+    }
+
+    private void writePObject(PObject object) throws IOException {
+      jsonWriter.beginObject();
+      var clazz = object.getClassInfo();
+      jsonWriter.name("moduleName").value(clazz.getModuleName());
+      jsonWriter.name("class").value(clazz.getDisplayName());
+      jsonWriter.name("moduleUri").value(clazz.getModuleUri().toString());
+      jsonWriter.name("properties");
+      jsonWriter.beginObject();
+      for (var kv : object.getProperties().entrySet()) {
+        jsonWriter.name(kv.getKey());
+        writeGenericObject(kv.getValue());
+      }
+      jsonWriter.endObject();
+      jsonWriter.endObject();
+    }
+
+    private void writeGenericObject(Object value) throws IOException {
+      if (value instanceof PNull) {
+        jsonWriter.nullValue();
+      } else if (value instanceof PObject po) {
+        writePObject(po);
+      } else if (value instanceof String string) {
+        jsonWriter.value(string);
+      } else if (value instanceof Boolean b) {
+        jsonWriter.value(b);
+      } else if (value instanceof Long l) {
+        jsonWriter.value(l);
+      } else if (value instanceof Double d) {
+        jsonWriter.value(d);
+      } else if (value instanceof List<?> list) {
+        jsonWriter.beginArray();
+        for (var v : list) {
+          writeGenericObject(v);
+        }
+        jsonWriter.endArray();
+      } else if (value instanceof Set<?> set) {
+        jsonWriter.beginArray();
+        for (var v : set) {
+          writeGenericObject(v);
+        }
+        jsonWriter.endArray();
+      } else if (value instanceof Map<?, ?> map) {
+        jsonWriter.beginObject();
+        for (var kv : map.entrySet()) {
+          var key = kv.getKey();
+          if (key instanceof String s) {
+            jsonWriter.name(s);
+          } else {
+            throw new PklException(
+                "Error serializing annotation for PklProject:\n"
+                    + "  cannot render map with non-string key: "
+                    + key);
+          }
+          writeGenericObject(kv.getValue());
+        }
+        jsonWriter.endObject();
+      } else if (value instanceof Pattern pattern) {
+        jsonWriter.value(pattern.pattern());
```

Comment:
Added support for the missing cases and disambiguated `Pattern`, `Set`, `Map`, etc.
I still don't support `PClass` and `TypeAlias` because these types are too complex (sometimes recursive) and I think it's fine to not support them for metadata.
Also added a test to roudtrip all new types.